compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUMaterializeEncodingPass.cpp

```diff
@@ -10,8 +10,8 @@
 
```

Comment:
This file also has some VMVX specific things. Maybe we can separate this out into `LLVMCPUMaterializeEncodingPass.cpp` and `VMVXMaterializeEncodingPass.cpp`. This will help with the WIP PRs from @bjacob

Reply:
I'll rebase the PR on top of @bjacob's work, so it won't affect his work. I'm thinking to separate this out into two files in a separate PR, so we wont touch the file a lot. It's cleaner for reviewer to me. What do you think?

Reply:
I am curious to see what the rebased PR looks like. Since the PR is based on a `origin` branch, you can rebase at any time without waiting for it to be merged.

What I would like to understand is: most of the code will remain the same between VMVX and LLVMCPU, as only a fraction of the code in `MaterializeEncodingPass.cpp` is backend-specific. Just as an example, what happens with `struct MaterializeFlowDispatchTensorLoadOp`, which has a screenful of nontrivial `matchAndRewrite` code that's shared across backends? This this PR, it is moving along with the rest of this file, to `LLVMCPU/`. But VMVX needs the same code - so does it need to duplicate it?

Does *some* code need to remain in `Common/` ?

Also, note that even before my PR #11640 , there is already a stub VMVX-dynamic-tile-sizes implementation in `MaterializeEncodingPass.cpp`. It's just that prior to PR #11640, the dynamic size values are just constant 1. But the VMVX code is still there and now it's moving along into `LLVMCPU/`.  That's to say that the need to think about where does the VMVX specific code goes exists already today, without waiting for #11640.

Reply:
+1 to removing the LLVMCPU <-> VMVX deps (VMVX Passes.cpp now pulls in an LLVMCPU pass) and to ensuring that we aren't forking too much code - a base pass that could be shared in Common/ with small variants in LLVMCPU/ and VMVX/, options on the pass to control the behavior changed by which pipeline was registering the pass, etc.

Reply:
(having some common behavior here will be even more important here when we want to call microkernels from codegen - we don't want to have that duplicated even if the same microkernels are being called in both LLVMCPU and VMVX)

Reply:
Interesting, I originally thought is about making distinguish between pack/unpack and pad/extract_slice approach, and thought that VMVX can depend on LLVMCPU.

Maybe we can have both implementation (pack/unpack and pad/extract_slice) to live in `Common/` and the backend can populate the patterns they want when creating their own passes. The target specific things seem can be carried by type converter, so we can create the CPU typeConverter, VMVX typeConvert, and llvmgpu typeConvert, etc.

I'll rethink about it, and create a new PR.

Reply:
This is along the lines of what I had in mind. Here was what I was thinking
1) The set of conversion patterns that work on upstream/linalgext operations live in LinalgExt.
2) The set of conversion patterns that work on flow/hal operations live in Codegen/Common.
3) The pass itself would need to be duplicated for each backend, so a LLVMCPUMaterializePass/VMVXCPUMaterializePass, etc. I think within the pass, only the callbacks should be different. I dont think we need a whole heirarchy of type converters. Am I missing something?

Reply:
I think we're aligned. I have a rough structure and can sort it out later today. I'm working on factoring out common things to `populate*` and will create a pass for each backend using the method. (I had a lot of meeting today..)

Reply:
I have a new draft version about this, please take a look at https://github.com/iree-org/iree/pull/11803